File: server/lib/funes/mcp/brief.ex
defmodule Server.MCP.Brief do
  @moduledoc """
  JSON-safe renderings of the reads the channel serves — one renderer per row
  type, shared by the `get_brief`/`get_facts` tools AND the resources: one
  source, two protocol doors. Every fact carries its read-time `certainty`
  (§4a: the certainty axis is a QUERY over provenance × check_cmd — `stated` >
  `checked` > `opinion`), so a brief can never launder an unchecked hunch into
  received truth (pi doc §2b). Caps keep their counts — a cut without a count
  lies.
  """

  alias Server.Event
  alias Server.Fact
  alias Server.Habit
  alias Server.Issue
  alias Server.Message
  alias Server.Question
  alias Server.Todo
  alias Server.Workspace

  @doc "A Board.brief/1 map, rendered."
  def scope(scope) do
    %{
      "thread_id" => scope.thread.id,
      "goal" => scope.goal,
      "todos" => capped(scope.todos, &todo/1),
      "next" => next(scope.next),
      "lead" => scope.lead,
      "done" => capped(scope.done, &done_entry/1),
      "learnings" => capped(scope.learnings, &fact/1),
      "unknowns" => capped(scope.unknowns, &question/1),
      "blockers" => capped(scope.blockers, &issue/1),
      "checks" => capped(scope.checks, &check/1),
      "recent" => Enum.map(scope.recent, &message/1)
    }
  end

  @doc "A %{shown, more} cap, rendered with its count intact."
  def capped(%{shown: shown, more: more}, render) do
    %{"shown" => Enum.map(shown, render), "more" => more}
  end

  @doc "A fact with its read-time certainty rank."
  def fact(%Fact{} = f) do
    %{
      "id" => f.id,
      "kind" => f.kind,
      "text" => f.text,
      "provenance" => f.provenance,
      "check_cmd" => f.check_cmd,
      "certainty" => certainty(f),
      "at" => at(f.created_at)
    }
  end

  @doc """
  The §4 total order, computed at read time and named in words: `stated` (the
  operator's own quoted words), `checked` (derived, carrying the command that
  re-runs it), `opinion` (derived and unverifiable — rank it last, re-verify
  before building on it).
  """
  def certainty(%Fact{provenance: "stated"}), do: "stated"
  def certainty(%Fact{check_cmd: nil}), do: "opinion"
  def certainty(%Fact{}), do: "checked"

  @doc "An event; `summary`/`evidence` surface from its human-readable detail."
  def event(%Event{} = e) do
    detail = e.detail || %{}

    %{
      "id" => e.id,
      "kind" => e.kind,
      "summary" => detail["summary"],
      "evidence" => detail["evidence"],
      "at" => at(e.created_at)
    }
  end

  @doc "A todo — a plan step. `done_at` present only when done (it opens NULL)."
  def todo(%Todo{} = t) do
    %{
      "id" => t.id,
      "text" => t.text,
      "done_at" => at(t.done_at),
      "at" => at(t.created_at)
    }
  end

  @doc "NEXT — the first open todo, or nil (derived, never stored)."
  def next(nil), do: nil
  def next(%Todo{} = t), do: todo(t)

  @doc """
  One entry of the merged DONE view, tagged by `source`: a completed `todo` or a
  `work_landed` event, each at its own timestamp — the merge, never a copy.
  """
  def done_entry(%{source: :todo, at: done_at, row: %Todo{} = t}) do
    %{"source" => "todo", "id" => t.id, "text" => t.text, "at" => at(done_at)}
  end

  def done_entry(%{source: :event, row: %Event{} = e}) do
    e |> event() |> Map.put("source", "event")
  end

  @doc "A question — an UNKNOWN. `resolution` is the answer once resolved (nil while open)."
  def question(%Question{} = q) do
    %{
      "id" => q.id,
      "text" => q.text,
      "state" => q.state,
      "resolution" => q.resolution,
      "at" => at(q.created_at)
    }
  end

  @doc "An issue — the stack's tracker (pi doc §4b)."
  def issue(%Issue{} = i) do
    %{
      "id" => i.id,
      "summary" => i.summary,
      "evidence" => i.evidence,
      "found_by" => i.found_by,
      "state" => i.state,
      "at" => at(i.created_at)
    }
  end

  @doc "A measured check — a `check_passed`/`check_failed` event, keyed on the real exit code."
  def check(%Event{kind: kind, created_at: created_at, detail: detail}) do
    d = detail || %{}

    %{
      "passed" => kind == "check_passed",
      "cmd" => d["cmd"],
      "exit" => d["exit"],
      "tail" => d["tail"],
      "at" => at(created_at)
    }
  end

  @doc "A message on the thread. A consult ASK (consult_id set, not a mirror) is flagged so a
  consumer can render it as a request for an answer, not ambient chatter."
  def message(%Message{} = m) do
    %{
      "id" => m.id,
      "author" => m.author,
      "body" => m.body,
      "reply_to" => m.reply_to,
      "at" => at(m.created_at),
      "consult" => m.consult_id != nil and m.mirrored == false
    }
  end

  @doc """
  A HABIT — how the agent should work with the operator. `state` distinguishes an approved,
  always-loaded preference from one still `pending` review; a cut without a count lies, and a
  pending habit is not an approved one.
  """
  def habit(%Habit{} = h) do
    %{
      "id" => h.id,
      "text" => h.text,
      "rationale" => h.rationale,
      "state" => h.state,
      "at" => at(h.created_at)
    }
  end

  @doc """
  A WORLD — a composition aleph reads to drive its picker/survey/spawn. Machine-
  global (no thread scope): `paths`/`roster`/`knobs` round-trip as the JSON the
  columns hold — a list of globs, a list of `{archetype,name,model?,knobs}` maps,
  and a free-form map.
  """
  def workspace(%Workspace{} = w) do
    %{
      "id" => w.id,
      "name" => w.name,
      "type" => w.type,
      "scope" => w.scope,
      "paths" => w.paths,
      "roster" => w.roster,
      "knobs" => w.knobs,
      "at" => at(w.created_at)
    }
  end

  defp at(nil), do: nil
  defp at(%DateTime{} = dt), do: DateTime.to_iso8601(dt)
end

File: adapters/pi/src/brief.ts
// The honest brief (pi doc §2b). A pure function: the server's get_dossier JSON in, the
// prompt text a fresh session wakes to out. It is pure so it is testable, and so the
// wiring (extension.ts) can stay a thin transport around it.
//
// Two anti-laundering rules live here, because this is the one place a re-brief could
// launder a guess into ground truth:
//   1. Every claim shows its provenance. A `derived` fact with no check_cmd renders AS
//      a hunch — never as flat truth — so generation N's guess cannot become generation
//      N+1's received truth with no one lying at any step.
//   2. The brief states its own staleness and its cuts. "Last banked 40m ago; work
//      since is unrecorded" (§6: unrecorded time is named), and every capped section
//      carries its +N more (a cut without a count is the empty-world lie in miniature).

// The get_dossier shape — the server's Server.MCP.Brief.scope/1 (one source, two protocol
// doors). Timestamps are ISO8601 strings or null; certainty is the read-time §4a rank.
export type Certainty = "stated" | "checked" | "opinion";

export interface Fact {
  id: number;
  kind: string;
  text: string;
  provenance: string;
  check_cmd: string | null;
  certainty: Certainty;
  at: string | null;
}

export interface Todo {
  id: number;
  text: string;
  done_at: string | null;
  at: string | null;
}

// One entry of the merged DONE view — a completed todo or a work_landed event, each at
// its own timestamp. Tagged by `source` so the render tells a finished step from a
// shipped outcome.
export type DoneEntry =
  | { source: "todo"; id: number; text: string; at: string | null }
  | {
      source: "event";
      id: number;
      kind: string;
      summary: string | null;
      evidence: string | null;
      at: string | null;
    };

export interface Blocker {
  id: number;
  summary: string;
  evidence: string | null;
  found_by: string | null;
  state: string;
  at: string | null;
}

export interface ChatterMessage {
  id: number;
  author: string;
  body: string;
  reply_to: number | null;
  at: string | null;
  // A consult ASK (consult_id set, not a mirror) — render it as a request for an answer.
  consult?: boolean;
}

export interface Question {
  id: number;
  text: string;
  state: string;
  resolution: string | null;
  at: string | null;
}

// A measured check — a command's real exit code, recorded as passed/failed. Not a
// self-report: the number is the truth.
export interface Check {
  passed: boolean;
  cmd: string | null;
  exit: number | null;
  tail: string | null;
  at: string | null;
}

export interface Capped<T> {
  shown: T[];
  more: number;
}

export interface Dossier {
  thread_id: number;
  north_star: string | null;
  lead: string | null;
  todos: Capped<Todo>;
  next: Todo | null;
  learnings: Capped<Fact>;
  unknowns: Capped<Question>;
  done: Capped<DoneEntry>;
  blockers: Capped<Blocker>;
  checks: Capped<Check>;
  // The thread's commits, joined by the `Tlon-Thread` trailer. Optional: a server from before
  // the join still renders.
  commits?: Capped<Commit>;
  chatter: ChatterMessage[];
}

export interface Commit {
  sha: string;
  subject: string;
  author: string;
  at: string;
}

// The operator's handle, so a stated fact reads as HIS words. The server trusts only this
// author for `stated` provenance; the render names him for the reader.
const OPERATOR = "andrew";

const CHATTER_TAIL = 5;

export function renderBrief(d: Dossier, now: Date = new Date(), model?: string): string {
  const lines: string[] = [];

  const goal = d.north_star ?? "(untitled thread)";
  lines.push(`# Brief: ${goal}`);
  lines.push(`Lead: ${d.lead ?? "unstaffed"}  ·  thread ${d.thread_id}`);
  // The session states who it is, so a model signs its commits with its OWN name and can't
  // copy a wrong one from an example. AGENTS.md "commit as who you are" reads from here.
  if (model) lines.push(`You are ${model} (pi).`);
  lines.push(staleness(d, now));

  if (d.todos.shown.length > 0) {
    lines.push("", "## Plan  (→ is NEXT, the first open step)");
    for (const t of d.todos.shown) {
      const isNext = d.next != null && t.id === d.next.id;
      lines.push(`${isNext ? "→" : " "} [ ] ${t.text}`);
    }
    pushMore(lines, d.todos.more, "todos");
  }

  if (d.learnings.shown.length > 0) {
    lines.push("", "## What's known");
    for (const f of d.learnings.shown) lines.push(renderFact(f));
    pushMore(lines, d.learnings.more, "learnings");
  }

  if (d.unknowns.shown.length > 0) {
    lines.push("", "## Unknowns  (open questions — resolve or you inherit the doubt)");
    for (const q of d.unknowns.shown) lines.push(`- ? ${q.text}`);
    pushMore(lines, d.unknowns.more, "unknowns");
  }

  if (d.blockers.shown.length > 0) {
    lines.push("", "## Blockers");
    for (const b of d.blockers.shown) lines.push(renderBlocker(b));
    pushMore(lines, d.blockers.more, "blockers");
  }

  if (d.checks.shown.length > 0) {
    lines.push("", "## Checks  (measured — ✓ passed, ✗ failed at that exit)");
    for (const c of d.checks.shown) lines.push(renderCheck(c));
    pushMore(lines, d.checks.more, "checks");
  }

  if (d.commits && d.commits.shown.length > 0) {
    lines.push("", "## Commits  (this thread's, in its repo)");
    for (const c of d.commits.shown) lines.push(`- ${c.sha} ${c.subject} — ${c.author}`);
    pushMore(lines, d.commits.more, "commits");
  }

  if (d.done.shown.length > 0) {
    lines.push("", "## Done");
    for (const e of d.done.shown) lines.push(renderDone(e));
    pushMore(lines, d.done.more, "done");
  }

  if (d.chatter.length > 0) {
    lines.push("", "## Recent chatter");
    for (const m of d.chatter.slice(-CHATTER_TAIL)) lines.push(renderChatter(m));
  }

  // A standing nudge, not state — the one behavior that keeps a thread from going cold. The
  // `coordinate-via-funes` skill teaches it, but a flash model reads a skill once and forgets;
  // the brief is injected every turn, so the reminder rides here too. Without it a rich session
  // banks nothing and its successor is briefed cold (§4c: the channel is the continuity
  // mechanism, not politeness).
  lines.push(
    "",
    "— Post as you work with `post_message` — the thread, not this session, is the memory a successor inherits; judgment you keep in your own context dies with the session.",
  );

  return lines.join("\n");
}

// A consult ask is a request for an answer, not ambient chatter — make the reply affordance
// explicit so the round-trip the bridge exists for actually fires.
function renderChatter(m: ChatterMessage): string {
  if (m.consult) return `- 📨 ${m.author} is consulting you — reply to answer: ${m.body}`;
  return `- ${m.author}: ${m.body}`;
}

// The default case FAILS CLOSED: `opinion`, or any certainty the server emits that we don't
// model, degrades to the least-authoritative rendering — never an unmarked line, which
// would launder an unknown into apparent truth.
function renderFact(f: Fact): string {
  switch (f.certainty) {
    case "stated":
      return `- [stated] ${OPERATOR}: "${f.text}"`;
    case "checked":
      return `- [checked] ${f.text}  (verify: \`${f.check_cmd}\`)`;
    default:
      return `- [hunch] ${f.text}  — unverified; re-check before you rely on it`;
  }
}

function renderBlocker(b: Blocker): string {
  const who = b.found_by ? `, found by ${b.found_by}` : "";
  const ev = b.evidence ? ` — ${b.evidence}` : "";
  return `- ${b.summary} (${b.state}${who})${ev}`;
}

function renderDone(e: DoneEntry): string {
  if (e.source === "todo") return `- [x] ${e.text}`;
  const ev = e.evidence ? ` — ${e.evidence}` : "";
  return `- ${e.summary ?? "(work landed)"}${ev}`;
}

function renderCheck(c: Check): string {
  const cmd = c.cmd ?? "(check)";
  return c.passed ? `- ✓ ${cmd}` : `- ✗ ${cmd} (exit ${c.exit ?? "?"})`;
}

function pushMore(lines: string[], more: number, noun: string): void {
  if (more > 0) lines.push(`_… and ${more} more ${noun} not shown (get_facts to read past the cap)._`);
}

// The staleness line: how old is the newest thing banked, and the honest warning that
// anything since is unrecorded. A thread with nothing banked says exactly that rather
// than rendering a confident empty world (§6: a cache that reports an empty world lies).
function staleness(d: Dossier, now: Date): string {
  const newest = newestAt(d);
  if (newest === null) {
    return "_Nothing has been banked on this thread yet — you are starting cold._";
  }
  return `_Last banked activity ${ago(newest, now)}; any work since is unrecorded._`;
}

function newestAt(d: Dossier): Date | null {
  const stamps: number[] = [];
  const collect = (at: string | null) => {
    if (!at) return;
    const t = Date.parse(at);
    if (!Number.isNaN(t)) stamps.push(t);
  };
  for (const f of d.learnings.shown) collect(f.at);
  for (const q of d.unknowns.shown) collect(q.at);
  for (const b of d.blockers.shown) collect(b.at);
  for (const c of d.checks.shown) collect(c.at);
  for (const e of d.done.shown) collect(e.at);
  for (const t of d.todos.shown) collect(t.at);
  for (const m of d.chatter) collect(m.at);
  if (stamps.length === 0) return null;
  return new Date(Math.max(...stamps));
}

// A stamp, never a clock (§4). Minutes up to 90, then hours, then days — enough for a
// reader to feel how stale the brief is without inventing precision.
function ago(then: Date, now: Date): string {
  const mins = Math.max(0, Math.round((now.getTime() - then.getTime()) / 60_000));
  if (mins < 90) return `${mins}m ago`;
  const hours = Math.round(mins / 60);
  if (hours < 48) return `${hours}h ago`;
  return `${Math.round(hours / 24)}d ago`;
}
